apis/policies/v1/clusteradmissionpolicy_webhook.go

```diff
@@ -84,7 +85,42 @@ func (r *ClusterAdmissionPolicy) ValidateUpdate(old runtime.Object) error {
 	return validatePolicyUpdate(oldPolicy, r)
 }
 
+func validatePolicyCreate(policy Policy) error {
+	return validateRulesField(policy)
+}
+
+// Validates the spec.Rules field for non-empty, webhook-generable rules
+func validateRulesField(policy Policy) error {
+	errs := field.ErrorList{}
+
+	if len(policy.GetRules()) != 0 {
+		rulesField := field.NewPath("spec", "rules")
+		for _, rule := range policy.GetRules() {
+			if len(rule.Operations) == 0 {
+				opField := rulesField.Child("operations")
+				errs = append(errs, field.Required(opField, "a value must be specified"))
+			} else if len(rule.Rule.APIGroups) == 0 || len(rule.Rule.APIVersions) == 0 || len(rule.Rule.Resources) == 0 {
```

Comment:
This works. I just tested it locally:

```console
kubectl delete ClusterAdmissionPolicy --all;                                                                                                                                            
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
No resources found
clusteradmissionpolicy.policies.kubewarden.io/privileged-pods created

```

However, we should block empty strings:

```console
kubectl delete ClusterAdmissionPolicy --all;                                                                                                                                              
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: [""]
    apiVersions: [""]
    resources: [""]    
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
clusteradmissionpolicy.policies.kubewarden.io "privileged-pods" deleted
clusteradmissionpolicy.policies.kubewarden.io/privileged-pods created
```

Empty lists are working fine:

```console
kubectl delete ClusterAdmissionPolicy --all;                                                                                                                                              
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: []  
    apiVersions: []  
    resources: []  
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
clusteradmissionpolicy.policies.kubewarden.io "privileged-pods" deleted
The ClusterAdmissionPolicy "privileged-pods" is invalid: spec.rules: Required value: at least one of apiGroups, apiVersions, or resources must have a specified value
```

Remember to add unit tests to cover these scenarios. :)

Reply:
This is working:

```console
kubectl delete ClusterAdmissionPolicy --all;                                                                                                                                          
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
No resources found
clusteradmissionpolicy.policies.kubewarden.io/privileged-pods created
```

It's working because the validation checks the list size only. This means that a policy with empty strings in all the fields will be accepted as well:

```console
kubectl delete ClusterAdmissionPolicy --all;
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: [""]
    apiVersions: [""]
    resources: [""]
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
clusteradmissionpolicy.policies.kubewarden.io "privileged-pods" deleted
clusteradmissionpolicy.policies.kubewarden.io/privileged-pods created
```

While empty list will be rejected:

```console
kubectl delete ClusterAdmissionPolicy --all;                                                                                                                                              
kubectl apply -f - <<EOF
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: privileged-pods
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.2.2
  rules:
  - apiGroups: []
    apiVersions: []
    resources: []
    operations:
    - CREATE
    - UPDATE
  mutating: false
EOF
clusteradmissionpolicy.policies.kubewarden.io "privileged-pods" deleted
The ClusterAdmissionPolicy "privileged-pods" is invalid: spec.rules: Required value: at least one of apiGroups, apiVersions, or resources must have a specified value
```

I think we should block when the fields (`apiGroups`,`apiVersions`,`resources`) have lists where all the entries are empty strings. 

I could not remember if we consider empty strings as valid default values. And I could not find something in the docs as well. So, I'm considering an issue now that we are validating the fields. Let me know if I'm missing something.

Reply:
@jvanz my bad, this scenario fully slipped my mind. I guess empty string is valid in `apiGroups` when it targets core objects, but empty string is invalid for both versions and resources. WDYT?

Reply:
Yes, I agree. This sounds the right path to me now.

Reply:
Updated!